File: 153_find_minimum_in_rotated_sorted_array.js
/**
 * @param {number[]} nums
 * @return {number}
 */

// iteratively

var findMin = function (nums) {
  if (nums.length === 1) return nums[0];
  let ref = nums[0];
  let min = 0;
  let max = nums.length - 1;

  while (true) {
    let midway = Math.floor((min + max) / 2);
    if (min >= max - 1) {
      return Math.min(nums[min], nums[max], ref);
    }

    // if number midway through is bigger than ref
    if (nums[midway] > ref) {
      if (nums[midway + 1] < ref) return nums[midway + 1];

      min = midway;
      continue;
    }

    // if number midway through is less than ref
    if (nums[midway - 1] > ref) return nums[midway];

    max = midway;

  }
};


// recursive attempt

// var findMin = function (nums) {
//   if (nums.length <= 1) return nums[0];
//   let ref = nums[0];
//   let midway = Math.floor(nums.length / 2);

//   if (nums[midway] > ref) {
//     if (nums[midway + 1] < ref) return nums[midway + 1];

//     return findMin(nums.slice(midway));
//   }

//   if (nums[midway - 1] < ref) return nums[midway - 1];

//   return findMin(nums.slice(0, midway));
// };

File: leetcode/153_find_minimum_in_rotated_sorted_array.py
class Solution:
    def findMin(self, nums) -> int:
        if nums[0] < nums[-1]:
            return nums[0]
        if len(nums) <= 3:
            return min(nums)
        lower_bound = 0
        upper_bound = len(nums) - 1
        while True:
            midway = (lower_bound + upper_bound) // 2
            if midway == lower_bound or midway == upper_bound:
                return min(nums[lower_bound], nums[upper_bound], nums[midway])
            if nums[midway] > nums[0]:
                if nums[midway + 1] < nums[0]:
                    return nums[midway + 1]
                lower_bound = midway
            else:
                if nums[midway - 1] > nums[0]:
                    return nums[midway]
                upper_bound = midway
